File: Arvores/ArvoreBinaria/Java/Recursivo/BinaryTree.java
package Arvores.ArvoreBinaria.Java.Recursivo;

public class BinaryTree<T extends Comparable<T>> {
    //======PRIVATE=====//
    private Node<T> root;
    private int length = 0;
    private int index = 0;

    //=====CONSTRUCTOR=====//
    public BinaryTree() {
        this.setRoot(null);
    }

    //=====GET=====//
    private Node<T> getRoot() {
        return this.root;
    }

    public int size() {
        return this.length;
    }

    //=====SET=====//
    private void setRoot(Node<T> root) {
        this.root = root;
    }

    private void inc() {
        this.length++;
    }

    //=====METODOS=====//
    //-----ADD-----//
    public void add(T obj) {
        this.add(obj, this.getRoot());
    }

    private void add(T obj, Node<T> node) {
        if(node != null) {
            if(obj.compareTo(node.getObj()) < 0) {
                if(node.getLeft() == null) {
                    node.setLeft(new Node<T>(obj));
                    this.inc();
                } else add(obj, node.getLeft());
            } else if(obj.compareTo(node.getObj()) > 0) {
                if(node.getRight() == null) {
                    node.setRight(new Node<T>(obj)); 
                    this.inc();
                } else add(obj, node.getRight());
            } else throw new InsertionError();
        } else {
            this.setRoot(new Node<T>(obj));
            this.inc();
        }
    }

    //-----INORDER-----//
    public void inOrder() {
        index = 0;
        System.out.print("[");
        this.inOrder(this.getRoot());
        System.out.println("]");
    }

    private void inOrder(Node<T> node) {
        if(node != null) {
            inOrder(node.getLeft());
            index++;
            System.out.print(node.getObj() + (index != this.size() ? ", " : ""));
            inOrder(node.getRight());
        }
    }

    //-----PREORDER-----//
    public void preOrder() {
        index = 0;
        System.out.print("[");
        this.preOrder(this.getRoot());
        System.out.println("]");
    }

    private void preOrder(Node<T> node) {
        if(node != null) {
            index++;
            System.out.print(node.getObj() + (index != this.size() ? ", " : ""));
            preOrder(node.getLeft());
            preOrder(node.getRight());
        }
    }

    //-----PREORDER-----//
    public void posOrder() {
        index = 0;
        System.out.print("[");
        this.posOrder(this.getRoot());
        System.out.println("]");
    }

    private void posOrder(Node<T> node) {
        if(node != null) {
            posOrder(node.getLeft());
            posOrder(node.getRight());
            index++;
            System.out.print(node.getObj() + (index != this.size() ? ", " : ""));
        }
    }

    //-----SEARCH-----//
    public Boolean search(T obj) {
        return this.search(obj, this.getRoot()).getObj() == obj;
    }

    private Node<T> search(T obj, Node<T> node) {
        if(node != null && node.getObj() != obj) {
            if(obj.compareTo(node.getObj()) < 0) node = search(obj, node.getLeft());
            else node = search(obj, node.getRight());
        } 
        
        return node;
    }

    //-----MAX-----//
    public T max() {
        return this.max(this.getRoot());
    }

    private T max(Node<T> node) {
        T content = null;
        if(node != null) {
            if(node.getRight() != null) content = max(node.getRight());
            else content = node.getObj();
        }
        return content;
    }

    //-----MIN-----//
    public T min() {
        return this.min(this.getRoot());
    }

    private T min(Node<T> node) {
        T content = null;
        if(node != null) {
            if(node.getLeft() != null) content = min(node.getLeft());
            else content = node.getObj();
        }
        return content;
    }

    //-----getHeigth-----//
    public int getHeight() {
        return this.getHeight(this.getRoot(), 0);
    }

    private int getHeight(Node<T> node, int h) {
        if(node == null) h--;
        else {
            int left = getHeight(node.getLeft(), h + 1);
            int right = getHeight(node.getRight(), h + 1);
            h = (left > right) ? left : right;
        }
        return h;
    }

    //=====OVERRIDE=====//
    @Override
    @SuppressWarnings("unchecked")
    public boolean equals(Object o) {
        return (this.getClass() == o.getClass()) ? compare(((BinaryTree<T>)o).getRoot(), this.getRoot()) : false;
    }

    private boolean compare(Node<T> node1, Node<T> node2) {
        Boolean value = node1 != null && node2 != null;
        if(value) {
            value = node1.getObj() == node2.getObj();
            value &= compare(node1.getLeft(), node2.getLeft());
            value &= compare(node1.getRight(), node2.getRight());
        } else if(node1 == null && node2 == null) value = true;
        return value;
    }

}

File: Arvores/ArvoreBinaria/C/Recursivo/BinaryTree.c

#include "header/BinaryTree.h"

//=====CONSTRUCTOR=====//
BinaryTree* construct() {
    BinaryTree* bst = (BinaryTree*)malloc(sizeof(BinaryTree));
    clear(bst);
    return bst;
}

//=====INITIALIZER=====//
void clear(BinaryTree* const bst) {
    bst->root = NULL;
    bst->len = 0;
}

//=====DESTRUCTOR=====//
void destructor(BinaryTree* const bst) {
    void destructorr(Node* const node);
    destructorr(bst->root);
    free(bst);
}

void destructorr(Node* const node) {
    if(node != NULL) {
        destructorr(node->left);
        destructorr(node->right);
        free(node);
    }
}

//=====METODOS=====//
//-----ISEMPTY-----//
bool isEmpty(BinaryTree* const bst) {
    return bst->root == NULL;
}

//-----ADD-----//
void add(BinaryTree* const bst, int obj) {
    void addr(BinaryTree* const, Node* const, int);

    if(bst->root != NULL) addr(bst, bst->root, obj);
    else bst->root = constructNodeObj(obj);
    bst->len++;
}

void addr(BinaryTree* const bst, Node* const node, int obj) {
    if(obj < node->obj) {
        if(node->left == NULL) node->left = constructNodeObj(obj);
        else addr(bst, node->left, obj);
    } else if(obj > node->obj) {
        if(node->right == NULL) node->right = constructNodeObj(obj);
        else addr(bst, node->right, obj);
    } else {
        printf("Insertion Error\n");
        bst->len--;
    }
}

//-----GETHEIGHT-----//
int getHeight(BinaryTree* const bst) {
    int getHeightr(Node* const, int);
    return getHeightr(bst->root, 0);
}

int getHeightr(Node* const node, int h) {
    if(node != NULL) {
        int left = getHeightr(node->left, h + 1);
        int right = getHeightr(node->right, h + 1);
        h = (left > right) ? left : right;
    }

    return h;
}

//-----INORDER-----//
void inOrder(BinaryTree* const bst) {
    int inOrderr(Node* const, int);
    printf("[");
    inOrderr(bst->root, bst->len);
    printf("]\n");
}

int inOrderr(Node* const node, int i) {
    if(node != NULL) {
        i = inOrderr(node->left, i);
        i--;
        printf("%i", node->obj);
        if(i != 0) printf(", ");
        i = inOrderr(node->right, i);
    }
    return i;
}

//-----PREORDER-----//
void preOrder(BinaryTree* const bst) {
    int preOrderr(Node* const, int);
    printf("[");
    preOrderr(bst->root, bst->len);
    printf("]\n");
}

int preOrderr(Node* const node, int i) {
    if(node != NULL) {
        i--;
        printf("%i", node->obj);
        if(i != 0) printf(", ");
        i = preOrderr(node->left, i);
        i = preOrderr(node->right, i);
    }
    return i;
}

//-----POSORDER-----//
void posOrder(BinaryTree* const bst) {
    int posOrderr(Node* const, int);
    printf("[");
    posOrderr(bst->root, bst->len);
    printf("]\n");
}

int posOrderr(Node* const node, int i) {
    if(node != NULL) {
        i = posOrderr(node->left, i);
        i = posOrderr(node->right, i);
        i--;
        printf("%i", node->obj);
        if(i != 0) printf(", ");
    }
    return i;
}

//-----SEARCH-----//
bool search(BinaryTree* const bst, int obj) {
    Node* searchr(Node* node, int obj);
    return searchr(bst->root, obj) == obj;
}

Node* searchr(Node* node, int obj) {
    if(node != NULL && node->obj != obj) {
        if(obj < node->obj) node = searchr(node->left, obj);
        else node = searchr(node->right, obj);
    }

    return node;
}

//-----MIN-----//
int min(BinaryTree* const bst) {
    int minr(Node* const);
    return (bst->root != NULL) ? minr(bst->root) : 0;
}

int minr(Node* const node) {
    int obj = 0;
    if(node->left != NULL) obj = minr(node->left);
    else obj = node->obj;
    return obj;
}

//-----MAX-----//
int max(BinaryTree* const bst) {
    int maxr(Node* const);
    return (bst->root != NULL) ? maxr(bst->root) : 0;
}

int maxr(Node* const node) {
    int obj = 0;
    if(node->left != NULL) obj = maxr(node->right);
    else obj = node->obj;
    return obj;
}
